src/stk/filters/decomposable_filter.inl

```diff
@@ -0,0 +1,140 @@
+#pragma once
+
+namespace
+{
+
+/*!
+ * \brief Auxiliary function.
+ */
+template<typename TVoxelType, typename TKernelType, typename TOutputType = TVoxelType>
+stk::VolumeHelper<TOutputType> decomposable_filter_3d(
+        const stk::VolumeHelper<TVoxelType>& img,
+        const stk::FilterKernel3<TKernelType>& kernel,
+        const stk::BorderMode border_mode
+        )
+{
+    stk::VolumeHelper<TOutputType> result(img.size(), TOutputType{0});
+    result.copy_meta_from(img);
+
+    const int3 dims = {int(img.size().x), int(img.size().y), int(img.size().z)};
+
+    // X dimension
+    #pragma omp parallel for
+    for (int y = 0; y < dims.y; ++y)
+    {
+        for (int z = 0; z < dims.z; ++z)
+        {
+            for (int x = 0; x < dims.x; ++x)
+            {
+                TOutputType value{0};
+                for (int t = -kernel.x.radius(); t <= kernel.x.radius(); t++)
+                {
+                    const TVoxelType v = img.at(x + t, y, z, border_mode);
+                    value += TOutputType(v * kernel.x[t]);
+                }
+                result(x, y, z) = value;
+            }
+        }
+    }
+
+    stk::VolumeHelper<TOutputType> tmp = result.clone();
+
+    //Y dimension
+    #pragma omp parallel for
+    for (int x = 0; x < dims.x; ++x)
+    {
+        for (int z = 0; z < dims.z; ++z)
+        {
+            for (int y = 0; y < dims.y; ++y)
+            {
+                TOutputType value{0};
+                for (int t = -kernel.y.radius(); t <= kernel.y.radius(); t++)
+                {
+                    const TOutputType v = tmp.at(x, y + t, z, border_mode);
+                    value += TOutputType(v * kernel.y[t]);
+                }
+                result(x, y, z) = value;
+            }
+        }
+    }
+
+    tmp.copy_from(result);
+    
+    //Z dimension
+    #pragma omp parallel for
+    for (int x = 0; x < dims.x; ++x)
+    {
+        for (int y = 0; y < dims.y; ++y)
+        {
+            for (int z = 0; z < dims.z; ++z)
+            {
+                TOutputType value{0};
+                for (int t = -kernel.z.radius(); t <= kernel.z.radius(); t++)
+                {
+                    const TOutputType v = tmp.at(x, y, z + t, border_mode);
+                    value += TOutputType(v * kernel.z[t]);
+                }
+                result(x, y, z) = value;
+            }
+        }
+    }
+    return result;
+}
+
+} // namespace
+
+
+namespace stk {
+
+template <typename TKernelType, typename TOutputType>
+stk::Volume decomposable_filter_3d(
+            const stk::Volume& volume,
+            const FilterKernel3<TKernelType> kernel,
+            const stk::BorderMode border_mode
+            )
+{
+    switch (volume.voxel_type())
```

Comment:
This is becoming a repeating pattern. I tried implementing a generic function to handle this dispatch w.r.t. the type of the volume in a reusable way, with some template sorcery like
```c++
template <typename TOut, template<typename> class fun, typename... TArgs>
TOut parametric(
        stk::Volume volume,
        TArgs &... args)
{
    switch (volume.voxel_type())
    {
    case stk::Type_Float:
        fun<float> f_float;
        return f_float(args...);
    case stk::Type_Double:
        fun<double> f_double;
        return f_double(args...);
    case stk::Type_Float3:
        fun<float3> f_float3;
        return f_float3(args...);
    default:
        FATAL() << "Unsupported voxel format";
    };
    throw std::runtime_error("This is a bug, the line should be unreachable");
}
```
that can be then used like
```c++
struct my_functor // this can be a template as well, if necessary
{
    template <typename T> // this is the type dispatched by `parametric`
    struct do_stuff_here {
        stk::Volume operator()(const stk::Volume& volume /* possibly other arguments */)
        {
            // ...
            return some_volume; /* the return type can also be something else */
        }
    };

    stk::Volume operator() (const stk::Volume& volume, /* same other arguments */)
    {
        return stk::parametric<stk::Volume, do_stuff_here, const stk::Volume& /* other args */>
                              (volume, volume, /* other args */);
    }
};
```
The problem is that this looks like a complete mess. Maybe it can be done without having to implement things as functors, but I am not 100% sure it is possible in C++17. I tried with just a templated function instead, but it did not work. An alternative is of course C variadic functions, but that would be no less horrible.